text/0000-mir.md

```diff
@@ -0,0 +1,821 @@
+  precise patterns in use. This is all vastly simplified and explicit
+  in MIR.
+  
+One thing the current MIR does not make explicit as explicit as it
+could is when something is *moved*. For by-value uses of a value, the
+code must still consult the type of the value to decide if that is a
+move or not. This could be made more explicit in the IR.
+
+### Which analyses are well-suited to the MIR?
+
+Some analyses are better suited to the AST than to a MIR. The
+following is a list of work the compiler does that would benefit from
+using a MIR:
+
+- **liveness checking**: this is used to issue warnings about unused assignments
+  and the like. The MIR is perfect for this sort of data-flow analysis.
+- **borrow and move checking**: the borrow checker already uses a
+  combination of the CFG and `ExprUseVisitor` to try and achieve a
+  similarly low-level of detail.
+- **translation to LLVM IR**: the MIR is much closer than the AST to
+  the desired end-product.
+  
+Some other passes would probably work equally well on the MIR or an
+AST, but they will likely find the MIR somewhat easier to work with
+than the current AST simply because it is, well, simpler:
+
+- **rvalue checking**, which checks that things are `Sized` which need to be.
+- **reachability** and **death checking**.
+
+These items are likely ill-suited to the MIR as designed:
+
+- **privacy checking**, since it relies on explicit knowledge of paths that is not
+  necessarily present in the MIR.
+- **lint checking**, since it is often dependent on the sort of surface details
+  we are seeking to obscure.
+
+For some passes, the impact is not entirely clear. In particular,
+**match exhaustiveness checking** could easily be subsumed by the MIR
+construction process, which must do a similar analysis during the
+lowering process. However, once the MIR is built, the match is
+completely desugared into more primitive switches and so forth, so we
+will need to leave some markers in order to know where to check for
+exhaustiveness and to reconstruct counter examples.
+
+# Detailed design
+
+### What is *really* being proposed here?
+
+The rest of this section goes into detail on a particular MIR design.
+However, the true purpose of this RFC is not to nail down every detail
+of the MIR -- which are expected to evolve and change over time anyway
+-- but rather to establish some high-level principles which drive the
+rest of the design:
+
+1. We should indeed lower the representation from an AST to something
+   else that will drive later analyses, and this representation should
+   be based on a CFG, not a tree.
+2. This representation should be explicitly minimal and not attempt to retain
+   the original syntactic structure, though it should be possible to recover enough
+   of it to make quality error messages.
+3. This representation should encode drops, panics, and other
+   scope-dependent items explicitly.
+4. This representation does not have to be well-typed Rust, though it
```

Comment:
well-typed Rust? The representation allows for some unsafe operations (e.g. unrestricted downcasts, unchecked indexing, calling unsafe functions) but *should* type-check.